### PR TITLE
adding functionality to allow evaluation of a single probe

### PR DESCRIPTION
**Pull request by karchern**
Hi Alessio,

since I'm working on more probe design problems, I've augmented your package by allowing the evaluation of a given k-mer directly.

### CODE REVIEW

**Comment by cschu** [comment]
@karchern Want me to review that?

**Comment by karchern** [comment]
I'd be very happy about that. I didn't take the time to kick out unnecessary loops etc (oriented myself mostly on Alessio's implementation).

File: fish_probes/C_predict_probes.py [hunk 1]
```diff
@@ -89,7 +89,53 @@ def find_conserved_regions(seq_sel_clade,k,perc_seq_with_kmer):
     # return
     return kmers_recall,kmers_precision
 
+def get_kmer_sens(seq_sel_clade,kmer):
```
Comment (cschu) [inline fish_probes/C_predict_probes.py]:
```suggestion
def get_kmer_sens(seq_sel_clade, kmer):
```
space after comma

File: fish_probes/C_predict_probes.py [hunk 2]
```diff
@@ -89,7 +89,53 @@ def find_conserved_regions(seq_sel_clade,k,perc_seq_with_kmer):
...
 
+    #all_strings_kmers = dict()
+    #for s in seq_sel_clade:
+    #   allK,allK_N = find_kmers(seq_sel_clade[s],len(kmer))
+    #   all_strings_kmers[s] = allK
+
+    all_kmers = set()
+    all_kmers.add(kmer)
```
Comment (cschu) [inline fish_probes/C_predict_probes.py]:
```suggestion
    all_kmers = {kmer}
```

File: fish_probes/C_predict_probes.py (hunk 1 above)
Comment (cschu) [inline fish_probes/C_predict_probes.py]:
```suggestion
def get_kmer_sens(seq_sel_clade,kmer):
    from collections import Counter
```

File: fish_probes/C_predict_probes.py [hunk 3]
```diff
@@ -89,7 +89,53 @@ def find_conserved_regions(seq_sel_clade,k,perc_seq_with_kmer):
...
+        UTIL_log.print_message("Calculating sensitivity for k-mer {}".format(kmer))
+    # # now we count how many times it appear
+    count_mers = dict()
+    for kmer in list(all_kmers):
+        count_mers[kmer] = 0
+    # # now we add the counts per k-mer
+    #for _, kmers in all_strings_kmers.items():
+    for _, seq in seq_sel_clade.items():
+        if kmer in seq:
+            count_mers[kmer] = count_mers[kmer] + 1
```
Comment (cschu) [inline fish_probes/C_predict_probes.py]: In the for-loop in line 114, you only ever check the last `kmer` from `list(all_kmers)`, not all. If you want to check all of `all_kmers`, then you'd have to nest the for-loop in 114 into the one in 110.

File: fish_probes/C_predict_probes.py [hunk 4]
```diff
@@ -89,7 +89,53 @@ def find_conserved_regions(seq_sel_clade,k,perc_seq_with_kmer):
...
+    #for s in all_strings_kmers:
+    #   for kmer in all_strings_kmers[s]:
+    #       all_kmers.add(kmer)
+
+    if VERBOSE > 2:
+        #UTIL_log.print_message("Identifed "+str(len(all_kmers))+" unique "+str(k)+"-mers.")
+        UTIL_log.print_message("Calculating sensitivity for k-mer {}".format(kmer))
+    # # now we count how many times it appear
+    count_mers = dict()
+    for kmer in list(all_kmers):
```
Comment (cschu) [inline fish_probes/C_predict_probes.py]:
```suggestion
    for kmer in all_kmers:
```
no need to convert, you can iterate over the set

File: fish_probes/C_predict_probes.py [hunk 5]
```diff
@@ -89,7 +89,53 @@ def find_conserved_regions(seq_sel_clade,k,perc_seq_with_kmer):
...
+        count_mers[kmer] = 0
+    # # now we add the counts per k-mer
+    #for _, kmers in all_strings_kmers.items():
+    for _, seq in seq_sel_clade.items():
```
Comment (cschu) [inline fish_probes/C_predict_probes.py]:
```suggestion
    for seq in seq_sel_clade.values():
```

File: fish_probes/C_predict_probes.py (hunk 3 above)
Comment (cschu) [inline fish_probes/C_predict_probes.py]:
```suggestion
            count_mers[kmer] += 1
```

File: fish_probes/C_predict_probes.py [hunk 6]
```diff
@@ -115,6 +161,22 @@ def check_uniqueness(kmers_precision, seq_other, probe_len):
         UTIL_log.print_message("The selected probes map to other "+str(n_matching_to_other)+" sequences.\n")
     return other_sel_clades
 
+def check_uniqueness_fast(kmers_precision, seq_other, probe_len):
+    n_matching_to_other = 0
+    # we check if the kmers are covered by other sequences
+    other_sel_clades = dict()
+    if len(kmers_precision.keys()) != 1:
```
Comment (cschu) [inline fish_probes/C_predict_probes.py]:
```suggestion
    if len(kmers_precision) != 1:
```
len(dict) == len(dict.keys())

File: fish_probes/C_predict_probes.py [hunk 7]
```diff
@@ -115,6 +161,22 @@ def check_uniqueness(kmers_precision, seq_other, probe_len):
...
+        asddassda
+    # is only on...
+    for kmer in kmers_precision.keys():
```
Comment (cschu) [inline fish_probes/C_predict_probes.py:139]:
```suggestion
    for kmer in kmers_precision:
```
no need to iterate over keys()

File: fish_probes/C_predict_probes.py [hunk 8]
```diff
@@ -115,6 +161,22 @@ def check_uniqueness(kmers_precision, seq_other, probe_len):
...
+        for s, seq in seq_other.items():
+            if kmer in seq:
+                kmers_precision[kmer] = kmers_precision[kmer] + 1
```
Comment (cschu) [inline fish_probes/C_predict_probes.py]:
```suggestion
                kmers_precision[kmer] += 1
```

File: fish_probes/C_predict_probes.py [hunk 9]
```diff
@@ -115,6 +161,22 @@ def check_uniqueness(kmers_precision, seq_other, probe_len):
...
+                if not kmer in other_sel_clades:
+                    other_sel_clades[kmer] = list()
+                other_sel_clades[kmer].append(s)
```
Comment (cschu) [inline fish_probes/C_predict_probes.py]:
```suggestion
                other_sel_clades.setdefault(kmer, []).append(s)
```

File: fish_probes/UTIL_arg_parser.py [hunk 10]
```diff
@@ -10,7 +10,7 @@ def input_parser(tool_version):
 
     # COMMAND
     parser.add_argument('command', action="store", default=None,
-                        help='command',choices=['design','check_probe','test'])
+                        help='command',choices=['design','check_probe','test','evaluate_probe_sens_spec'])
```
Comment (cschu) [inline fish_probes/UTIL_arg_parser.py]:
```suggestion
                        help='command', choices=['design', 'check_probe', 'test', 'evaluate_probe_sens_spec'])
```

File: fish_probes/C_predict_probes.py (hunk 3 above)
Comment (cschu) [reply fish_probes/C_predict_probes.py]: This is the crucial review item - the rest is cosmetics and low-level tuning.

**Review by cschu — CHANGES_REQUESTED** [review]
Check `fish_probes/C_predict_probes:110-116` in particular - that's a functional issue, the rest is minor.

File: fish_probes/UTIL_arg_parser.py (hunk 10 above)
Comment (cschu) [reply fish_probes/UTIL_arg_parser.py]: @karchern You can review and accept/reject my suggestions. Best to go to the "Files changed" tab and add accepted suggestions to batch.

File: fish_probes/C_predict_probes.py (hunk 3 above)
Comment (karchern) [reply fish_probes/C_predict_probes.py]: Thanks Christian. This is very unelegant but not incorrect: In this function, `all_kmers` will only ever contain one entry (see line 99, 100 in `fish_probes/C_predict_probes.py`)

I'm comitting a fix that makes this more clear.

File: fish_probes/C_predict_probes.py (hunk 2 above)
Comment (karchern) [reply fish_probes/C_predict_probes.py]: I'm going to remove this

File: fish_probes/C_predict_probes.py (hunk 4 above)
Comment (karchern) [reply fish_probes/C_predict_probes.py]: thanks!

**Comment by karchern** [comment]
Hi @cschu, I think this can be merged now. Would be great if it could be merged since Fabian would like to use this functionality.

Cheers,
Nic

**Pull request merged**